Site tutorial: Part 1 (#85)
davenverse/shellfish: docs/tutorial/reading_writing.md

# Reading and writing

In this section, we will guide you through the process of opening and extracting information from files using our simple, intuitive API. Whether you are working with text, binary, or other formats, you will find the tools you need to seamlessly integrate file reading into your Scala project. Let us dive in and unlock the power of data stored in files!

> **Important:** To avoid tutorial hell, we recommend doing a really small project while you read this tutorial to place acquired knowledge into practice. It does not have to be fancy, but it should tap into your curiosity and challenge you just a little.

Reviewer: ```suggestion
```
Personal opinion: while I think this is great advice, I don't think this is the place to put it.

Reviewer 2: +1

Author: Maybe on the introduction section?

@:choice(syntax)

```scala
import shellfish.Path
import shellfish.syntax.path.*
```

@:choice(static)

```scala
import shellfish.Path
import shellfish.FilesOs
```

@:@

Next, define the `run` function to execute the `IO`. To do that, you need to extend your application with a `cats.effect.IOApp`, let us name it `App`:

```scala mdoc:compile-only
import cats.effect.{IO, IOApp}

object App extends IOApp.Simple:

  def run: IO[Unit] = ???

end App
```

Now we can start using the library! First, create a `Path` containing the path to the file you want to read:

```scala mdoc:compile-only
import fs2.io.file.Path // TODO: delete this and use the Shellfish's one
val path = Path("testdata/readme.txt")
```

And use the `read` function to load the file in memory as a string:

@:select(api-style)

@:choice(syntax)

```scala mdoc:fail
import cats.effect.{IO, IOApp}

object App extends IOApp.Simple:

  val path = Path("testdata/readme.txt")

  def run: IO[Unit] = path.read

end App
```

@:choice(static)

```scala mdoc:fail
import cats.effect.{IO, IOApp}

object App extends IOApp.Simple:

  val path = Path("testdata/readme.txt")

  def run: IO[Unit] = FilesOs.read(path)

end App
```

@:@

> Oops! We got an error saying that the `run` function accepts an `IO[Unit]`, but the `read` function returns an `IO[String]`. This happens because we are not doing anything with the string loaded and therefore not returning `IO[Unit]`. Fortunately, because `IO` is a monad, we can sequence the value obtained by evaluating the `read` computation with another that prints things to the console (and thus returns `IO[Unit]`!). This is achieved by using the `flatMap` function as follows:

Reviewer: ```suggestion
Oops! We got an error saying that the `run` function accepts an `IO[Unit]`, but the `read` function returns an `IO[String]`. This happens because we are not doing anything with the string and therefore not returning `IO[Unit]`. Let's fix that by sequencing the value obtained from `read` with another that prints it to the console (and thus returns `IO[Unit]`). This can be achieved using the `flatMap` function as follows:
```

Scary  m-word

@:select(api-style)

@:choice(syntax)

```scala
path.read.flatMap(file => IO(println(file)))
```

@:choice(static)

```scala
FilesOs.read(path).flatMap(file => IO(println(file)))
```

@:@

> As you might know, what is happening above is that we are calling the `flatMap` method and passing as parameter a function describing what we want to do with the `file` inside the `IO`, in this case, passing it to the computation `IO(println(file))`.

Reviewer: ```suggestion
What is happening above is that we are calling the `flatMap` method and passing as parameter a function describing what we want to do with the `file` inside the `IO`, in this case, passing it to the computation `IO(println(file))`.
```

Now pass the program to the `run` method and everything should go nicely:

@:select(api-style)

@:choice(syntax)

```scala mdoc:compile-only
import io.chrisdavenport.shellfish.syntax.path.*
import fs2.io.file.Path
import cats.effect.{IO, IOApp}

object App extends IOApp.Simple:

  val path = Path("testdata/readme.txt")

  def run: IO[Unit] = path.read.flatMap(file => IO(println(file)))

end App
```

@:choice(static)

```scala mdoc:compile-only
import io.chrisdavenport.shellfish.FilesOs
import fs2.io.file.Path
import cats.effect.{IO, IOApp}

object App extends IOApp.Simple:

  val path = Path("testdata/readme.txt")

  def run: IO[Unit] = FilesOs.read(path).flatMap(file => IO(println(file)))

end App
```

@:@

Congratulations! You have just loaded the contents of a file in pure Scala 🎉.

### Exercise

> You may not like to keep using the `flatMap` function over and over again to sequence computations. This is why there is the [macro keyword `for`](https://docs.scala-lang.org/scala3/book/control-structures.html#for-expressions) to automatically let the compiler write the `flatMap`s for you. Why don't you try rewriting the program we just did using for-comprehensions? 

Reviewer: ```suggestion
You may not like to keep using the `flatMap` function over and over again to sequence computations. This is why there is the [`for` construct](https://docs.scala-lang.org/scala3/book/control-structures.html#for-expressions) to automatically let the compiler write the `flatMap`s for you. Why don't you try rewriting the program we just did using for-comprehensions? 
```

Reviewer 2: +1000. `for` is definitely not a macro but a compiler built in feature. It's embedded in the language, not an external integration.

Author: I thought I had changed it before :P I must have forgotten to commit it. Thanks for bringing it back!

```scala
def run: IO[Unit] =
  for
    // Complete your code here!
    ...
  yield ()
```
[See solution](https://gist.github.com/f8cf952fd5f7341ef698dfe825cfca93.git)

## Writing and modifying the contents of a file

Now that you know how to load a file, you might also want to modify it and save it.

> To write to a file, use the `write` function with one of its variants so that you can save it to a completely different file. Here, we reverse the file so it reads backwards (if that is something you want to do):

Reviewer: ```suggestion
To write to a file, use the `write` function to save the contents to a new file. Here, we reverse the file so it reads backwards:
```

@:select(api-style)

@:choice(syntax)

```scala mdoc:compile-only
import io.chrisdavenport.shellfish.syntax.path.*
import fs2.io.file.Path
import cats.effect.{IO, IOApp}

object App extends IOApp.Simple:

  val path = Path("testdata/change_me.txt")

  def run: IO[Unit] =
    for
      file <- path.read
      reversedFile = file.reverse
      _ <- path.write(reversedFile)
    yield ()

end App
```

@:choice(static)

```scala mdoc:compile-only
import io.chrisdavenport.shellfish.FilesOs
import fs2.io.file.Path
import cats.effect.{IO, IOApp}

object App extends IOApp.Simple:

  val path = Path("testdata/change_me.txt")

  def run: IO[Unit] =
    for
      file <- FilesOs.read(path)
      reversedFile = file.reverse
      _ <- FilesOs.write(path, reversedFile)
    yield ()

end App
```

@:@

Be aware that this will overwrite the contents of the file. So be careful not to change important files while you are learning!

### Exercise

Try loading the contents of two different files, concatenating them, and saving the result to a third location. How would you do it?

[See possible solution](https://gist.github.com/99880af9b051b15f8adace7fe7aee2e3.git)

## Working line by line

> There is a useful variant of the standalone methods called `Lines` (e.g. `readLines`), which reads the file line by line and stores them on a `List[String]`. This comes handy when you are working with a list of things that you want to convert:

Reviewer: ```suggestion
Shellfish provides a method called `readLines`, which reads the file line by line and stores them on a `List[String]`. This comes handy when you are working with a list of things that you want to convert:
```
the sentence reads a bit confusing to me, standalone methods of what? I don't think you have to mention there are variants, if `readLines` will be the default.

Author: I mention that on the wiki section :)

`testdata/names.data`

```
Alex
Jamie
Morgan
Riley
Taylor
Casey
River
```

@:select(api-style)

@:choice(syntax)

```scala mdoc:compile-only
import io.chrisdavenport.shellfish.syntax.path.*
import fs2.io.file.Path
import cats.effect.{IO, IOApp}

object Names extends IOApp.Simple:

  case class Name(value: String)

  val namesPath = Path("testdata/names.data")

  def run: IO[Unit] =
    for
      lines <- namesPath.readLines      // (1)
      names = lines.map(Name(_))        // (2)
      _ <- IO(names.foreach(println))   // (3)
    yield ()

end Names
```

@:choice(static)

```scala mdoc:compile-only
import io.chrisdavenport.shellfish.FilesOs
import fs2.io.file.Path
import cats.effect.{IO, IOApp}

object Names extends IOApp.Simple:

  case class Name(value: String)

  val namesPath = Path("testdata/names.data")

  def run: IO[Unit] =
    for
      lines <- FilesOs.readLines(namesPath)   // (1)
      names = lines.map(Name(_))              // (2)
      _ <- IO(names.foreach(println))         // (3)
    yield ()

end Names
```

@:@

> In the example from above, we first load the list of names as a `List` of `String` (1), then we map over this list and convert the strings to a `List` of `Name` (2), and finally we print each of one the elements recently mapped to the console (3). Note that using `println` with `foreach` creates side effects, that's why we wrap the computation on an `IO`.

Reviewer: So when you have numbered items like this you'll want to structure the footnotes like:
```
(1) load the list of names as `List`
(2) convert to `Name`
(3) print
```
This is easier to parse (imo!!) for most people. However I think this is a nit, and very subjective so feel free to completely ignore this lol

Reviewer 2: +1

### Exercise

> Imagine we want to add some spacing between lines on a text and save the result in a different file:

Reviewer: ```suggestion
Write a function that reads a file into lines, adds a blank line between the lines and saves the result in a different file. For example given the following input:
```

_`testdata/edgar_allan_poe/no_spaced_dream.txt`_

```
Take this kiss upon the brow!
And, in parting from you now,
Thus much let me avow —
You are not wrong, who deem
That my days have been a dream;
Yet if hope has flown away
In a night, or in a day,
In a vision, or in none,
Is it therefore the less gone?
All that we see or seem
Is but a dream within a dream.
```

> Convert it to:

Reviewer: ```suggestion
Your function should output the following:
```
